Framework/Algorithms/test/PolarizationCorrections/HeliumAnalyserEfficiencyTest.h

```diff
@@ -0,0 +1,280 @@
+  void testSampleFit() {
+    // For a p of e.g. p=0.1, we're going to generate some spin flipped and non-spin flippped
+    // data and fit to it. We should recover the original p.
+    const double p = 0.1;
+    const double pxd = 12;
+    const double mu = pxd * 0.0733;
+    const double te = 1;
+
+    const boost::format nsfFunc = boost::format("0.5*%1%*exp(-x*%2%*(1-%3%))") % te % mu % p;
+    const boost::format sfFunc = boost::format("0.5*%1%*exp(-x*%2%*(1+%3%))") % te % mu % p;
+
+    auto nsfWs1 = generateFunctionDefinedWorkspace("nsfWs1", nsfFunc.str());
+    auto nsfWs2 = generateFunctionDefinedWorkspace("nsfWs2", nsfFunc.str());
+    auto sfWs1 = generateFunctionDefinedWorkspace("sfWs1", sfFunc.str());
+    auto sfWs2 = generateFunctionDefinedWorkspace("sfWs2", sfFunc.str());
+    auto grpWs = groupWorkspaces("grpWs", {nsfWs1, sfWs1, nsfWs2, sfWs2});
+
+    auto heliumAnalyserEfficiency = AlgorithmManager::Instance().create("HeliumAnalyserEfficiency");
+    heliumAnalyserEfficiency->initialize();
+    heliumAnalyserEfficiency->setProperty("InputWorkspace", grpWs->getName());
+    heliumAnalyserEfficiency->setProperty("SpinStates", "11,10,00,01");
+    heliumAnalyserEfficiency->setProperty("TransmissionEmptyCell", te);
+    heliumAnalyserEfficiency->setProperty("GasPressureTimesCellLength", pxd);
+    heliumAnalyserEfficiency->setProperty("HeliumPolarisation", "p_He");
+    heliumAnalyserEfficiency->execute();
+
+    MatrixWorkspace_sptr pHeWs = AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>("p_He");
+    auto pFromFit = pHeWs->y(0)[0];
+    TS_ASSERT_DELTA(pFromFit, p, 1e-5);
+  }
+
+  void testCovarianceMatrix() {
+    // Should accept a 2x2 covariance matrix for the two input parameters
+    auto wsGrp = createExampleGroupWorkspace("wsGrp");
+
+    ITableWorkspace_sptr covariance = WorkspaceFactory::Instance().createTable();
+
+    covariance->addColumn("str", "Name");
+    covariance->addColumn("double", "a");
+    covariance->addColumn("double", "b");
+    covariance->appendRow();
+
+    // Test covariance matrix with incorrect size
+    auto heliumAnalyserEfficiency = AlgorithmManager::Instance().create("HeliumAnalyserEfficiency");
+    heliumAnalyserEfficiency->initialize();
+    heliumAnalyserEfficiency->setProperty("InputWorkspace", wsGrp->getName());
+    heliumAnalyserEfficiency->setProperty("Covariance", covariance);
+    TS_ASSERT_THROWS(heliumAnalyserEfficiency->execute(), std::runtime_error &);
+
+    // Test with correctly sized covariance matrix
+    covariance->appendRow();
+    for (size_t i = 0; i < 2; ++i)
+      for (size_t j = 1; j < 3; ++j) {
+        covariance->cell<double>(i, j) = 5;
+      }
+
+    heliumAnalyserEfficiency->initialize();
+    heliumAnalyserEfficiency->setProperty("InputWorkspace", wsGrp->getName());
+    heliumAnalyserEfficiency->setProperty("Covariance", covariance);
+    heliumAnalyserEfficiency->execute();
+    TS_ASSERT_EQUALS(true, heliumAnalyserEfficiency->isExecuted());
```

Comment:
Is there anything we can add here that would test that the covariance matrix has been used correctly by the algorithm (i.e. any data values that could be checked)? I think at the moment if a regression was introduced into how the algorithm uses the covariance matrix then this test would still pass.

Reply:
I've added a couple of tests that check the error output with different covariance matrices.